### PR TITLE
prevent joining an already existed response topic

### PR DESCRIPTION
**Pull request by merlinran**
With re-publishing, this can be seen in bidbot frequently. This is a fix.
```
2021-08-04T16:37:33.488+0200  ERROR  mpeer/pubsub  pubsub/pubsub.go:283  creating response topic: joining topic: topic already exists
2021-08-04T16:38:17.039+0200  ERROR  mpeer/pubsub  pubsub/pubsub.go:283  creating response topic: joining topic: topic already exists
2021-08-04T16:38:46.664+0200  ERROR  mpeer/pubsub  pubsub/pubsub.go:283  creating response topic: joining topic: topic already exists
2021-08-04T16:40:21.257+0200  ERROR  mpeer/pubsub  pubsub/pubsub.go:283  creating response topic: joining topic: topic already exists
```

### CODE REVIEW

File: rpc.go [hunk 1]
```diff
@@ -309,7 +309,10 @@ func (t *Topic) listen() {
 
 func (t *Topic) publishResponse(from peer.ID, id cid.Cid, data []byte, e error) {
 	topic, err := newTopic(t.ctx, t.ps, t.host, responseTopic(t.t.String(), from), false)
-	if err != nil {
+	if err != nil &&
+		// this can happen when more than one messages are received
+		// from the same peer.
+		!strings.Contains(err.Error(), "topic already exists") {
```
[the remark this answers is not in the document]
Comment (merlinran) [reply rpc.go:317]: A decision also needs to be made here: Should we continue or simply return if topic exists? For our usage, an existing topic means duplicated message, which should simply be ignored (the only topic with more than one messages is the AuctionTopic, which doesn't watch for response), but for general user case, it should be totally fine to send many messages on a single topic and expect many responses.

I vote for correctness of this package and let the caller handle duplications

[the remark this answers is not in the document]
Comment (merlinran) [reply rpc.go:317]: Oop continuing won't work... as `top` is `nil` for whatever `err` returned.

**Comment by merlinran** [comment]
There's some previous discussion around this behaviour of `Join` https://github.com/libp2p/go-libp2p-pubsub/issues/253

File: rpc.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (sanderpick) [reply rpc.go:317]: this must happen when the first response topic hasn't been closed yet. I think it's reasonable to assume only one response can be sent at a time... but many can be sent over the lifetime of the topic.

conclusion: just ignore the error and return from `publishResponse`.

[the remark this answers is not in the document]
Comment (merlinran) [reply rpc.go:317]: Again, this relates to race condition - because `publishResponse` runs in a separate goroutine, what if there are actually more than one messages arrive, hence start several goroutines? All but the first one would be responded.

Even if we respond all messages, they can be out of order, hence a bigger problem: the other end gets response for a wrong message. The simplest would be to stop spawning new goroutine for response.

File: rpc.go [hunk 2]
```diff
@@ -296,10 +299,8 @@ func (t *Topic) listen() {
 			data, err := handler(msg.ReceivedFrom, t.t.String(), msg.Data)
 			if !strings.Contains(t.t.String(), "/_response") {
 				// This is a normal message; respond with data and error
-				go func() {
-					msgID := cid.NewCidV1(cid.Raw, util.Hash(msg.Data))
-					t.publishResponse(msg.ReceivedFrom, msgID, data, err)
-				}()
+				msgID := cid.NewCidV1(cid.Raw, util.Hash(msg.Data))
+				t.publishResponse(msg.ReceivedFrom, msgID, data, err)
```
Comment (merlinran) [inline rpc.go:303]: Ended up making `publishResponse` synchronous. It is actually not a blocking call as I anticipated. What it does: 1) call `newTopic`, which returns immediately, and 2) call `topic.Publish`, which blocks when calling `t.p.disc.Bootstrap` https://github.com/libp2p/go-libp2p-pubsub/blob/master/topic.go#L245, but only if `WithReadiness` is set, which we don't. So, synchronous call always close the response topic, and responses are never out of order.

File: rpc.go [hunk 3]
```diff
@@ -91,6 +91,7 @@ func NewTopic(ctx context.Context, ps *pubsub.PubSub, host peer.ID, topic string
 	}
 	t.resTopic, err = newTopic(ctx, ps, host, responseTopic(topic, host), true)
 	if err != nil {
+		t.cancel()
```
Comment (merlinran) [inline rpc.go:94]: drive-by. cleaning up goroutines in case of error.

**Comment by merlinran** [comment]
@sanderpick how do you feel about the changes now?

File: rpc.go (hunk 2 above)
Comment (sanderpick) [reply rpc.go:303]: Nice, that sounds much better

**Review by sanderpick — APPROVED** [review]
LGTM! 💪

File: rpc.go (hunk 3 above)
Comment (sanderpick) [reply rpc.go:94]: should this be `t.Close()`?

Comment (merlinran) [reply rpc.go]: Right... fixed. also added some lines to properly close resources in `newTopic` https://github.com/textileio/go-libp2p-pubsub-rpc/pull/13/commits/ec807b7e2f4383141a1aa3d3da0f1fcbe6c6e5d4

**Comment by merlinran** [comment]
Merging. Thanks for the discussion here!
